js/genkit/src/plugin.ts

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Genkit } from './genkit.js';
+
+export interface PluginProvider {
+  name: string;
+  initializer: () => void | Promise<void>;
+}
+
+type PluginInit = (genkit: Genkit) => void | Promise<void>;
+
+export type GenkitPlugin = (genkit: Genkit) => PluginProvider;
+
+/**
+ * Defines a Genkit plugin.
+ */
+export function genkitPlugin<T extends PluginInit>(
```

Comment:
Even though it isn't defining an action itself, I could see this as being renamed `definePlugin` ?

The initialiser is basically going to define a bunch of actions right

Reply:
Interesting... the only thing is that it follows separate mechanic compared to other `ai.define*` methods. When you call `ai.define*` it registers whatever you define with the genkit instance. However `definePlugin` is not doing that -- it only creates a plugin, but then there's a separate `ai.registerPlugin`  process (not exactly that, but conceptually) which makes the plugin available to the genkit instance.
I'm not sure whether anyone would care about that distinction... but to me it feels important enough. :)

Reply:
that's fair :D